connectors/tests/test_byoc.py

```diff
@@ -641,31 +712,127 @@ class Client:
     }
 
     await connector.prepare(config)
-    assert connector.source_klass.__doc__ == "Banana"
     assert connector.status == Status.NEEDS_CONFIGURATION
+    assert connector.service_type == service_type
+    assert not connector.configuration.is_empty()
+
+
+@pytest.mark.asyncio
+async def test_connector_validate_filtering_not_edited(patch_logger):
+    index = Mock()
+    index.update = AsyncMock()
+    validator = Mock()
+    validator.validate_filtering = AsyncMock()
+
+    connector = Connector(elastic_index=index, doc_source=DOC_SOURCE)
+    await connector.validate_filtering(validator=validator)
+
+    validator.validate_filtering.assert_not_awaited()
+    index.update.assert_not_awaited()
+
+
+@pytest.mark.asyncio
+async def test_connector_validate_filtering_invalid(patch_logger):
+    index = Mock()
+    index.update = AsyncMock()
+    validation_result = FilteringValidationResult(
+        state=FilteringValidationState.INVALID,
+        errors=[FilterValidationError(ids=[1], messages="something wrong")],
+    )
+    validator = Mock()
+    validator.validate_filtering = AsyncMock(return_value=validation_result)
+    expected_validation_update_doc = {
+        "filtering": [
+            {
+                "domain": DEFAULT_DOMAIN,
+                "draft": DRAFT_FILTERING_DEFAULT_DOMAIN
+                | {"validation": validation_result.to_dict()},
+                "active": ACTIVE_FILTERING_DEFAULT_DOMAIN,
+            },
+            FILTERING_OTHER_DOMAIN,
+        ]
+    }
+
+    connector = Connector(
+        elastic_index=index, doc_source=DOC_SOURCE_WITH_EDITED_FILTERING
+    )
+    await connector.validate_filtering(validator=validator)
+
+    validator.validate_filtering.assert_awaited()
+    index.update.assert_awaited_with(
+        doc_id=CONNECTOR_ID, doc=expected_validation_update_doc
+    )
+
+
+@pytest.mark.asyncio
+async def test_connector_validate_filtering_valid(patch_logger):
+    index = Mock()
+    index.update = AsyncMock()
+    validation_result = FilteringValidationResult()
+    validator = Mock()
+    validator.validate_filtering = AsyncMock(return_value=validation_result)
+    expected_validation_update_doc = {
+        "filtering": [
+            {
+                "domain": DEFAULT_DOMAIN,
+                "draft": DRAFT_FILTERING_DEFAULT_DOMAIN
+                | {"validation": validation_result.to_dict()},
+                "active": DRAFT_FILTERING_DEFAULT_DOMAIN
+                | {"validation": validation_result.to_dict()},
+            },
+            FILTERING_OTHER_DOMAIN,
+        ]
+    }
+
+    connector = Connector(
+        elastic_index=index, doc_source=DOC_SOURCE_WITH_EDITED_FILTERING
+    )
+    index.fetch_by_id = AsyncMock(return_value=connector)
+    await connector.validate_filtering(validator=validator)
+
+    validator.validate_filtering.assert_awaited()
+    index.update.assert_awaited_with(
+        doc_id=CONNECTOR_ID, doc=expected_validation_update_doc
+    )
```

Comment:
Weird test failure.

For these two tests, if I run them separately, they all pass. if I run
```
bin/pytest -sv connectors/tests/test_byoc.py
```
or 
```
make test
```

It randomly fails the second one:
```shell
_________________________________________________________________________________________________________________ test_connector_validate_filtering_invalid __________________________________________________________________________________________________________________

patch_logger = <connectors.conftest.Logger object at 0x106828e80>

    @pytest.mark.asyncio
    async def test_connector_validate_filtering_invalid(patch_logger):
        index = Mock()
        index.update = AsyncMock()
        validation_result = FilteringValidationResult(
            state=FilteringValidationState.INVALID,
            errors=[FilterValidationError(ids=[1], messages="something wrong")],
        )
        validator = Mock()
        validator.validate_filtering = AsyncMock(return_value=validation_result)
        expected_validation_update_doc = {
            "filtering": [
                {
                    "domain": DEFAULT_DOMAIN,
                    "draft": DRAFT_FILTERING_DEFAULT_DOMAIN
                    | {"validation": validation_result.to_dict()},
                    "active": ACTIVE_FILTERING_DEFAULT_DOMAIN,
                },
                FILTERING_OTHER_DOMAIN,
            ]
        }
    
        connector = Connector(
            elastic_index=index, doc_source=DOC_SOURCE_WITH_EDITED_FILTERING
        )
        await connector.validate_filtering(validator=validator)
    
>       validator.validate_filtering.assert_awaited()

connectors/tests/test_byoc.py:800: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <AsyncMock name='mock.validate_filtering' id='4407170128'>

    def assert_awaited(self):
        """
        Assert that the mock was awaited at least once.
        """
        if self.await_count == 0:
            msg = f"Expected {self._mock_name or 'mock'} to have been awaited."
>           raise AssertionError(msg)
E           AssertionError: Expected validate_filtering to have been awaited.

/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/mock.py:2240: AssertionError
```

I added some logs, and confirm that `validator` and `validator.validate_filtering` in two tests are different objects.

I'm still digging, but in case you encountered the same scenario before, cc @tarekziade

Reply:
Finally made it work. Here is the root cause:

Both tests use the same doc source `DOC_SOURCE_WITH_EDITED_FILTERING` to instantiate the connector:
https://github.com/elastic/connectors-python/blob/008f48ffa9f4e86481f75bc0d9f952f1462d405c/connectors/tests/test_byoc.py#L795-L797

https://github.com/elastic/connectors-python/blob/008f48ffa9f4e86481f75bc0d9f952f1462d405c/connectors/tests/test_byoc.py#L826-L828

when the first test runs, the method `validate_filtering` will try to update the data source here:
https://github.com/elastic/connectors-python/blob/54096a4afcedf4ffd923d4f1d499e78d983bc3e1/connectors/byoc.py#L687

which also changes `DOC_SOURCE_WITH_EDITED_FILTERING` state to either `valid` or `invalid`. When the 2nd test runs, it's not edited, thus it's not validated.

The fix is to add `deepcopy` in each test:
https://github.com/elastic/connectors-python/blob/54096a4afcedf4ffd923d4f1d499e78d983bc3e1/connectors/tests/test_byoc.py#L795-L797

https://github.com/elastic/connectors-python/blob/54096a4afcedf4ffd923d4f1d499e78d983bc3e1/connectors/tests/test_byoc.py#L826-L828

Reply:
I also ran into similar errors as I tend to forget that upper case only indicates something being constant without actually making it constant. Should we use the type hint `Final` (https://docs.python.org/3/library/typing.html#typing.Final) for constants in tests? We could add `mypy` to `make test` and fail when a constant is mutated. cc @tarekziade

Reply:
That opens a can of worms. 

- if we introduce type annotations we're going to dramatically change how we work. But I think that's worth a discussion
- `mypy` is a heavy dependency, if we can avoid it... maybe Google's `pytype` would work for checking this. Or Microsoft's `PyRight` -- to be tested